js/packages/openinference-instrumentation-langchain/src/utils.ts

```diff
@@ -0,0 +1,348 @@
+  const normalizedLangchainMessageClass = isString(langchainMessageClass)
+    ? langchainMessageClass.toLowerCase()
+    : "";
+  if (normalizedLangchainMessageClass.includes("human")) {
+    return "user";
+  }
+  if (normalizedLangchainMessageClass.includes("ai")) {
+    return "assistant";
+  }
+  if (normalizedLangchainMessageClass.includes("system")) {
+    return "system";
+  }
+  if (normalizedLangchainMessageClass.includes("function")) {
+    return "function";
+  }
+  if (
+    normalizedLangchainMessageClass.includes("chat") &&
+    isObject(messageData.kwargs) &&
+    isString(messageData.kwargs.role)
+  ) {
+    return messageData.kwargs.role;
+  }
+  return null;
+}
+
+/**
+ * Gets the content of a message from the langchain message kwargs.
+ * @param messageKwargs The langchain message kwargs to extract the content from
+ * @returns The content of the message or null
+ */
+function getContentFromMessageData(
+  messageKwargs: Record<string, unknown>,
+): string | null {
+  return isString(messageKwargs.content) ? messageKwargs.content : null;
+}
+
+function getFunctionCallDataFromAdditionalKwargs(
+  additionalKwargs: Record<string, unknown>,
+): LLMMessageFunctionCall {
+  const functionCall = additionalKwargs.function_call;
+  if (!isObject(functionCall)) {
+    return {};
+  }
+  const functionCallName = isString(functionCall.name)
+    ? functionCall.name
+    : undefined;
+  const functionCallArgs = isString(functionCall.args)
+    ? functionCall.args
+    : undefined;
+  return {
+    [SemanticConventions.MESSAGE_FUNCTION_CALL_NAME]: functionCallName,
+    [SemanticConventions.MESSAGE_FUNCTION_CALL_ARGUMENTS_JSON]:
+      functionCallArgs,
+  };
+}
+
+/**
+ * Gets the tool calls from the langchain message additional kwargs and formats them into OpenInference attributes.
+ * @param additionalKwargs The langchain message additional kwargs to extract the tool calls from
+ * @returns the OpenInference attributes for the tool calls
+ */
+function getToolCallDataFromAdditionalKwargs(
+  additionalKwargs: Record<string, unknown>,
+): LLMMessageToolCalls {
+  const toolCalls = additionalKwargs.tool_calls;
+  if (!Array.isArray(toolCalls)) {
+    return {};
+  }
+  const formattedToolCalls = toolCalls.map((toolCall) => {
+    if (!isObject(toolCall) && !isObject(toolCall.function)) {
+      return {};
+    }
+    const toolCallName = isString(toolCall.function.name)
+      ? toolCall.function.name
+      : undefined;
+    const toolCallArgs = isString(toolCall.function.arguments)
+      ? toolCall.function.arguments
+      : undefined;
+    return {
+      [SemanticConventions.TOOL_CALL_FUNCTION_NAME]: toolCallName,
+      [SemanticConventions.TOOL_CALL_FUNCTION_ARGUMENTS_JSON]: toolCallArgs,
+    };
+  });
+  return {
+    [SemanticConventions.MESSAGE_TOOL_CALLS]: formattedToolCalls,
+  };
+}
+
+/**
+ * Parses a langchain message into OpenInference attributes.
+ * @param messageData The langchain message data to parse
```

Comment:
minor nit. Best to try to follow true JSDoc if possible

```
/**
 * Represents a book.
 * @constructor
 * @param {string} title - The title of the book.
 * @param {string} author - The author of the book.
 */

Reply:
ahh yeah i removed the types cause they weren't showing up but can add them in, also i see the - in teh spec, but it also gets added automatically on hover in vscode so i was a bit confused, will add